### PR TITLE
Initial Fault TMDS register handler changes

### PR DESCRIPTION
**Pull request by mye956**
### Summary
This PR will add the initial fault tmds handler changes. We will create and register 3 types of HTTP request endpoints for each of the network fault injection types: `PUT(Start)`, `DELETE(Stop)`, and `GET(check status)`. In a future PR, we will implement the actual functionality of each handlers of these endpoints but for this PR, it will only take care of registering each of the endpoints and their handlers to an HTTP server router (will be the TMDS server router which will also be changed in a future PR).

### Implementation details
* `agent/handlers/task_server_setup.go`
  * New function called `registerFaultHandlers` which will take in the mux router of TMDS (will be done in a future PR), agent state, and metrics factory that will add new handlers for each of the network fault injection types to the passed in mux router
* `ecs-agent/tmds/handlers/fis/vi/handlers/handlers.go`
  * New struct type called `FaultHandler` to keep track of agent state and metrics factory
  * New function called `FaultNetworkFaultPath` that will take in a fault type and then return the formatted TMDS endpoint path for the fault. This function will be used in `RegisterFaultHandlers`
  * Start/Stop/Check network fault handler functions for each of the network fault types. Currently not implemented but will be done in a future PR
* `ecs-agent/tmds/handlers/fis/vi/types/types.go`
  * New constants called `BlackHolePortFaultType`, `LatencyFaultType`, and `PacketLossFaultType`

### Testing
* Added new unit tests that will test:
  * Getting the correct endpoint for each fault type via `FaultNetworkFaultPath`
  * Calling each PUT/DELETE/GET request for each of the fault types after calling `registerFaultHandlers`
  * Try calling an invalid endpoint after calling `registerFaultHandlers`

New tests cover the changes: Yes

### Description for the changelog
Feature - Adding initial fault TMDS register handler implementation 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+func RegisterFISHandlers(muxRouter *mux.Router, agentState state.AgentState, metricsFactory metrics.EntryFactory) {
+	handler := handler{
+		agentState:     agentState,
+		metricsFactory: metricsFactory,
+	}
+
+	// Setting up handler endpoints for network blackhole port fault injections
+	muxRouter.HandleFunc(
+		FISNetworkFaultPath(types.BlackHolePortFaultType),
+		handler.StartNetworkBlackholePortHandler(),
```
Comment (xxx0624) [inline ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go:47]: nit: `handler.StartNetworkBlackholePort()` should be good IMO. The second Handler looks not necessary.

File: ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+		handler.CheckPacketLossStatusHandler(),
+	).Methods("GET")
+
+	logger.Debug("FIS Handlers have been set up")
+
+}
+
+// FISNetworkFaultPath will take in a fault type and return the TMDS endpoint path
+func FISNetworkFaultPath(fault string) string {
+	return fmt.Sprintf("/api/%s/fis/v1/%s",
```
Comment (xxx0624) [inline ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go:92]: `fis` ->  `fault`

**Review by xxx0624 — APPROVED** [review]

File: ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+type handler struct {
+	// TODO: Mutex will be used in a future PR
+	// mu             sync.Mutex
+	agentState     state.AgentState
+	metricsFactory metrics.EntryFactory
+}
+
+// RegisterFISHandlers will initialize all of the Start/Stop/Check fault injection endpoints and register them to the TMDS router
+func RegisterFISHandlers(muxRouter *mux.Router, agentState state.AgentState, metricsFactory metrics.EntryFactory) {
```
Comment (singholt) [inline ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go:38]: registering handlers for the agent API v1 should be done by the respective agents right? (and not part of the shared library code here)?

for example, the docker agent registers agent api v1 handlers [here](https://github.com/aws/amazon-ecs-agent/blob/5f35c935eed13826121315dbf797b8d7b83f849d/agent/handlers/task_server_setup.go#L164)

File: ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+type handler struct {
+	// TODO: Mutex will be used in a future PR
+	// mu             sync.Mutex
+	agentState     state.AgentState
+	metricsFactory metrics.EntryFactory
+}
+
+// RegisterFISHandlers will initialize all of the Start/Stop/Check fault injection endpoints and register them to the TMDS router
```
Comment (amogh09) [inline ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go:37]: Let's not include "FIS" in any names.

Comment (mye956) [reply ecs-agent/tmds/handlers/fis/v1/handlers/handlers.go]: sure, will rename

File: agent/handlers/task_server_setup_test.go [hunk 5]
```diff
@@ -3561,3 +3566,114 @@ func TestUpdateTaskProtection(t *testing.T) {
 		},
 	}))
 }
+
+func TestRegisterHandler(t *testing.T) {
```
Comment (singholt) [inline agent/handlers/task_server_setup_test.go:3570]: TestRegisterFaultHandler

**Review by singholt — APPROVED** [review]

File: agent/handlers/task_server_setup_test.go (hunk 5 above)
Comment (mye956) [reply agent/handlers/task_server_setup_test.go:3570]: Ah yeah that's fair, will address in a follow up PR. thanks

File: agent/handlers/task_server_setup_test.go [hunk 6]
```diff
@@ -3561,3 +3566,114 @@ func TestUpdateTaskProtection(t *testing.T) {
...
+
+			registerFaultHandlers(router, agentState, metricsFactory)
+			var requestBody io.Reader
+			req, err := http.NewRequest(tc.method, fmt.Sprintf("/api/%s/fault/v1/%s", endpointId, tc.fault),
+				requestBody)
+			require.NoError(t, err)
+
+			// Send the request and record the response
+			recorder := httptest.NewRecorder()
+			router.ServeHTTP(recorder, req)
```
Comment (amogh09) [inline agent/handlers/task_server_setup_test.go:3674]: How will we update this test when the handlers are implemented?

Comment (mye956) [reply agent/handlers/task_server_setup_test.go:3674]: Good question, I was thinking of introducing different tests functions for each of the request types (i.e. one for Start blackhole port, Stop blackhole port, and check blackhole port) in future PRs. This test was more so of actually testing whether or not the handlers can be registered to the router. I can create separate test cases in this PR if that makes more sense.

**Review by amogh09 — APPROVED** [review]

**Pull request merged**